src/ai/backend/agent/resources.py

```diff
@@ -523,6 +523,53 @@ async def _wrap_future(fut: asyncio.Future) -> None:
     return slot_allocs
 
 
+async def scan_gpu_alloc_map(
+    kernel_ids: Sequence[KernelId],
+    scratch_root: Path,
+) -> dict[DeviceId, Decimal]:
+    """
+    Fetch the current allocated amounts for fractional gpu from
+    ``/home/config/resource.txt`` files in the kernel containers managed by this agent.
+    """
+
+    async def _read_kernel_resource_spec(path: Path) -> dict[DeviceId, Decimal]:
+        alloc_map: dict[DeviceId, Decimal] = defaultdict(lambda: Decimal(0))
+        try:
+            loop = asyncio.get_running_loop()
+            content = await loop.run_in_executor(None, path.read_text)
+            resource_spec = KernelResourceSpec.read_from_string(content)
+
+            if cuda := resource_spec.allocations.get(DeviceName("cuda")):
+                if cuda_shares := cuda.get(SlotName("cuda.shares")):
+                    for device_id, shares in cuda_shares.items():
+                        alloc_map[device_id] += Decimal(shares)
+
+                if cuda_device := cuda.get(SlotName("cuda.device")):
+                    for device_id, device in cuda_device.items():
+                        alloc_map[device_id] += Decimal(device)
+
+        except FileNotFoundError:
+            # there may be races with container destruction
+            return {}
+
+        return alloc_map
+
+    tasks = []
+    for kernel_id in kernel_ids:
+        path = scratch_root / str(kernel_id) / "config" / "resource.txt"
+        task = asyncio.create_task(_read_kernel_resource_spec(path))
+        tasks.append(task)
+
```

Comment:
TODO: Use TaskGroup instead of `asyncio.create_task`, and use proper constant size instead of `kernels_ids` length